spec/models/sample_manifest/uploader_spec.rb

```diff
@@ -302,5 +304,48 @@
       uploader.run!
       expect(uploader).to be_processed
     end
+
+    context 'with a pools plate manifest' do
+      let(:uploader) { described_class.new(test_file, SampleManifestExcel.configuration, user, false) }
+
+      before do
+        # create a test manifest file with 2 plates, 2 wells per plate, and 2 rows per well
+        download =
+          build(
+            :test_download_plates,
+            num_rows_per_well: 2,
+            columns: SampleManifestExcel.configuration.columns.pools_plate.dup
+          )
+        download.save(test_file_name)
+        Delayed::Worker.delay_jobs = false
+        uploader.run!
+      end
+
+      it 'will upload the manifest' do
+        expect(uploader).to be_processed
+      end
+
+      it 'will understand there is one pool per well' do
+        expect(uploader.upload.sample_manifest.pools).not_to be_nil
+        expect(uploader.upload.sample_manifest.pools.count).to eq(4) # one pool per well
+      end
+
+      it 'will set tag_depth on aliquots' do
+        labware = uploader.upload.sample_manifest.labware
+        expect(labware.count).to eq(2)
+
+        labware.each do |plate|
+          wells = plate.wells
+          expect(wells.count).to eq(2)
+
+          wells.each do |well|
+            aliquots = well.aliquots
+
+            # within a well, each aliquot should have its own tag_depth
+            expect(aliquots.map(&:tag_depth).uniq.count).to eq(aliquots.count)
+          end
+        end
+      end
+    end
```

Comment:
Maybe add tests for where a file is uploaded and some and/or all of the rows for a particular well are not filled in, to check that it works ok and correctly filters out the blanks without losing the rows that do have info.
e.g. num_rows_per_well 3, and A1 has 3 rows, but only 2 are filled in. or A2 has 3 rows and none are filled in.
(Or maybe this is already tested elsewhere as more basic functionality of filtering out empty rows first before we get here?)

Reply:
OK I've had a look and I think it's alright. There's a test in this file `'will upload a valid partial plate sample manifest'` which only fills out some of the rows. There's also unit testing of the specific bit of code that ignores empty rows - the method called `empty?` in `row.rb` is tested in `row_spec.rb`, test called `'knows if it is empty'`. Think that's enough as I didn't have to add any additional functionality around this.